File: Cosmos-Android/app/src/main/java/wannabit/io/cosmostaion/model/kava/CollateralParam.java
package wannabit.io.cosmostaion.model.kava;

import android.text.TextUtils;

import com.google.gson.annotations.SerializedName;

import java.math.BigDecimal;

import wannabit.io.cosmostaion.model.type.Coin;

public class CollateralParam {
    @SerializedName("denom")
    public String denom;

    @SerializedName("type")
    public String type;

    @SerializedName("liquidation_ratio")
    public String liquidation_ratio;

    @SerializedName("debt_limit")
    public Coin debt_limit;

    @SerializedName("stability_fee")
    public String stability_fee;

    @SerializedName("auction_size")
    public String auction_size;

    @SerializedName("liquidation_penalty")
    public String liquidation_penalty;

    @SerializedName("prefix")
    public int prefix;

    @SerializedName("conversion_factor")
    public String conversion_factor;

    @SerializedName("spot_market_id")
    public String spot_market_id;

    @SerializedName("liquidation_market_id")
    public String liquidation_market_id;

    public String getDpMarketId() {
        if (!TextUtils.isEmpty(denom) && !TextUtils.isEmpty(debt_limit.denom)) {
            return denom.toUpperCase() + " : " + debt_limit.denom.toUpperCase();
        }
        return "";
    }

    public BigDecimal getDpLiquidationRatio() {
        return new BigDecimal(liquidation_ratio).movePointRight(2);
    }

    public BigDecimal getDpLiquidationPenalty() {
        return new BigDecimal(liquidation_penalty).movePointRight(2);
    }

    public BigDecimal getDpStabilityFee() {
        return (new BigDecimal(stability_fee).subtract(BigDecimal.ONE)).multiply(new BigDecimal("31536000")).movePointRight(2);
    }

    public String getImagePath() {
        if (!TextUtils.isEmpty(spot_market_id))
            return spot_market_id.replace(":","")   +".png";
        return "";
    }
}

File: Cosmos-IOS/Cosmostation/Cosmostation/Model/KavaModel/CollateralParam.swift
//
//  CollateralParam.swift
//  Cosmostation
//
//  Created by 정용주 on 2021/03/04.
//  Copyright © 2021 wannabit. All rights reserved.
//

import Foundation

public struct CollateralParam {
    var denom: String?
    var type: String?
    var liquidation_ratio: String?
    var debt_limit: Coin?
    var stability_fee: String?
    var auction_size: String?
    var liquidation_penalty: String?
    var prefix: Int?
    var conversion_factor: String?
    var spot_market_id: String?
    var liquidation_market_id: String?
    
    init(_ dictionary: NSDictionary?) {
        self.denom = dictionary?["denom"] as? String
        self.type = dictionary?["type"] as? String
        self.liquidation_ratio = dictionary?["liquidation_ratio"] as? String
        if let rawDebtLimit = dictionary?["debt_limit"] as? NSDictionary {
            self.debt_limit = Coin.init(rawDebtLimit)
        }
        self.stability_fee = dictionary?["stability_fee"] as? String
        self.auction_size = dictionary?["auction_size"] as? String
        self.liquidation_penalty = dictionary?["liquidation_penalty"] as? String
        self.prefix = dictionary?["prefix"] as? Int
        self.conversion_factor = dictionary?["conversion_factor"] as? String
        self.spot_market_id = dictionary?["spot_market_id"] as? String
        self.liquidation_market_id = dictionary?["liquidation_market_id"] as? String
    }
    
    func getpDenom() -> String? {
        return debt_limit?.denom
    }
    
    func getcDenom() -> String? {
        return denom
    }
    
    func getMarketImgPath() -> String? {
        return denom! + "usd"
    }
    
    func getDpMarketId() -> String? {
        if let dpDenom = denom?.uppercased(), let dpDebtLimit = debt_limit?.denom.uppercased() {
            return dpDenom + " : " + dpDebtLimit
        }
        return ""
    }
    
    func getLiquidationRatio() -> NSDecimalNumber {
        return NSDecimalNumber.init(string: liquidation_ratio)
    }
    
    func getDpLiquidationRatio() -> NSDecimalNumber {
        return NSDecimalNumber.init(string: liquidation_ratio).multiplying(byPowerOf10: 2, withBehavior: WUtils.handler2Down)
    }
    
    func getDpLiquidationPenalty() -> NSDecimalNumber {
        return NSDecimalNumber.init(string: liquidation_penalty).multiplying(byPowerOf10: 2, withBehavior: WUtils.handler2Down)
    }
    
    func getDpStabilityFee() -> NSDecimalNumber {
        return (NSDecimalNumber.init(string: stability_fee).subtracting(NSDecimalNumber.one)).multiplying(by: NSDecimalNumber.init(string: "31536000")).multiplying(byPowerOf10: 2, withBehavior: WUtils.handler2Down)
    }
}
